udp_driver/src/udp_component.cpp

```diff
@@ -0,0 +1,69 @@
+// Copyright 2021 Apex.AI, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include "udp_component/udp_component.hpp"
+
+#include <thread>
+#include <vector>
+
+#include "rclcpp/rclcpp.hpp"
+
+
+namespace udp_component
+{
+
+UdpComponent::UdpComponent(const rclcpp::NodeOptions & options)
+: UdpDriverT("udp_component", options)
+{
+  RCLCPP_INFO(this->get_logger(), "Initializing udp_component");
+
+  // set output using already declared ROS parameters
+  this->get_parameter("ip", address_);
+  this->get_parameter("port", port_);
+  RCLCPP_INFO(this->get_logger(), "ip: %s", address_.c_str());
+  RCLCPP_INFO(this->get_logger(), "port: %i", port_);
+
+  RCLCPP_INFO(this->get_logger(), "Starting udp reader thread");
+  // start receiving packets on separate thread
+  reader_thread = std::thread{[this]() {this->run();}};
```

Comment:
Why not just pass this as a reference to the member function instead of creating a lambda?

```
reader_thread = std::thread(&UdpDriverNode::run, this);
```

Reply:
i think you meant to use `UdpComponent` above....but I already tried that earlier and posted above that it results in a compile error.

```
  required from here
/usr/include/c++/9/thread:120:44: error: static assertion failed: std::thread arguments must be invocable after conversion to rvalues
  120 |           typename decay<_Args>::type...>::value,
```

Reply:
so this doesnt throw an error when I add an argument, like this:
```
reader_thread = std::thread(&UdpDriverNode::run, this, 0U);
```

does this cause any negative sideeffects by initializing the `run` function with its default value? by not passing an argument to `run` it defaults to `0U` anyways, right?